README.md

```diff
@@ -1,52 +1,202 @@
+    - safety of the proposed change is checked.
+1. The configuration is rendered in to a set of kubernetes objects (e.g. DaemonSets).
+1. The desired objects are reconciled against the API server, being created or updated as necessary.
+1. The applied configuration is stored separately, for later comparison.
+
+### Filling defaults
+Because most of the operator's configuration parameters are not changeable, it is important that the applied configuration is stable across upgrades. This has two implications:
+
+**All defaults must be made explicit.**
+
+The network configuration is transformed internally in to a fully-expressed struct. All optional values must have their default set. For example, if the vxlanPort is not specified, the default of 4789 is chosen and applied to the OpenShiftSDNConfig.
+
+Making all defaults explicit makes it possible to prevent unsafe changes when a newer version of the operator changes a default value.
+
+**Some values must be carried forward from the previous configuration.**
+
+Some values are computed at run-time but can never be changed. For example, the MTU of the overlay network is determined from the node's interfaces. Changing this is unsafe, so this must always be carried forward.
+
+Note that the fully-expressed configuration is not persisted back to the apiserver. Instead, it is saved only in the stored applied configuration. An alternative would be to inject these values via a mutating webhook. That requires a running service network, which we don't have until after we've run.
+
+### Validation
+Each network provider is responsible for validating their view of the configuration. For example, the OpenShiftSDN provider validates that the vxlanPort is a number between 1 and 65535, that the MTU is sane, etc. No validation is provided via the CRD itself.
+
+## Building
+
+### Test builds
+Build binaries manually with
+```
+./hack/build-go.sh
+```
+
+There is a special mode, the renderer, that emulates what the operator _would_ apply, given a fresh cluster. You can execute it manually:
+
+```
+_output/linux/amd64/cluster-network-renderer --config sample-config.yaml --out out.yaml
+```
+
+## Running manually against a test cluster
+If you have a running cluster, you can run the operator manually. Just set the `KUBECONFIG` environment variable.
+
+Note that, like all operators installed by the Cluster Version Operator, it determines all pointers to dependent images via environment variables. You will have to set several environment varables to emulate how the CVO works.
+
+You can determine the current values of these environment variables by inspecting the valid deployed daemonset. Do this before you delete it...
+```sh
+kubectl get -n openshift-cluster-network-operator daemonset cluster-network-operator -ojson | jq -Mrj '.spec.template.spec.containers[0].env[] | select(has("value")) | @text "\(.name)=\(.value) "' | tee images
```

Comment:
you can do that with just kubectl and JSON Template...

also, should we be using `oc` rather than `kubectl` in this doc? Not sure what our style guide would say...

Reply:
update: I just needed to actually think for a minute, and the jsonpath stuff worked great. And we should use OC.